src/coefficients.jl

```diff
@@ -0,0 +1,699 @@
+    matrix_nnz::Int = 0
+
+    matrix_range::Vector{Float64} = sizehint!(Float64[1.0, 1.0], 2)
+    bounds_range::Vector{Float64} = sizehint!(Float64[1.0, 1.0], 2)
+    rhs_range::Vector{Float64} = sizehint!(Float64[1.0, 1.0], 2)
+    objective_range::Vector{Float64} = sizehint!(Float64[1.0, 1.0], 2)
+
+    variables_in_constraints::Set{VariableRef} = Set{VariableRef}()
+    variables_not_in_constraints::Vector{VariableRef} = VariableRef[]
+
+    empty_rows::Vector{ConstraintRef} = ConstraintRef[]
+    bound_rows::Vector{ConstraintRef} = ConstraintRef[]
+    dense_rows::Vector{Tuple{ConstraintRef,Int}} = Tuple{ConstraintRef,Int}[]
+
+    matrix_small::Vector{Tuple{ConstraintRef,VariableRef,Float64}} =
+        Tuple{ConstraintRef,VariableRef,Float64}[]
+    matrix_large::Vector{Tuple{ConstraintRef,VariableRef,Float64}} =
+        Tuple{ConstraintRef,VariableRef,Float64}[]
+
+    bounds_small::Vector{Tuple{VariableRef,Float64}} =
+        Tuple{VariableRef,Float64}[]
+    bounds_large::Vector{Tuple{VariableRef,Float64}} =
+        Tuple{VariableRef,Float64}[]
+
+    rhs_small::Vector{Tuple{ConstraintRef,Float64}} =
+        Tuple{ConstraintRef,Float64}[]
+    rhs_large::Vector{Tuple{ConstraintRef,Float64}} =
+        Tuple{ConstraintRef,Float64}[]
+
+    objective_small::Vector{Tuple{VariableRef,Float64}} =
+        Tuple{VariableRef,Float64}[]
+    objective_large::Vector{Tuple{VariableRef,Float64}} =
+        Tuple{VariableRef,Float64}[]
+
+    has_quadratic_objective::Bool = false
+    has_quadratic_constraints::Bool = false
+    sense::JuMP.OptimizationSense = JuMP.FEASIBILITY_SENSE
+
+    objective_quadratic_range = sizehint!(Float64[1.0, 1.0], 2)
+    matrix_quadratic_range = sizehint!(Float64[1.0, 1.0], 2)
+
+    nonconvex_objective::Bool = false
+    nonconvex_rows::Vector{ConstraintRef} = ConstraintRef[]
+
+    matrix_quadratic_small::Vector{
+        Tuple{ConstraintRef,VariableRef,VariableRef,Float64},
+    } = Tuple{ConstraintRef,VariableRef,VariableRef,Float64}[]
+    matrix_quadratic_large::Vector{
+        Tuple{ConstraintRef,VariableRef,VariableRef,Float64},
+    } = Tuple{ConstraintRef,VariableRef,VariableRef,Float64}[]
+    objective_quadratic_small::Vector{Tuple{VariableRef,VariableRef,Float64}} =
+        Tuple{VariableRef,VariableRef,Float64}[]
+    objective_quadratic_large::Vector{Tuple{VariableRef,VariableRef,Float64}} =
+        Tuple{VariableRef,VariableRef,Float64}[]
+
+end
+
+function _update_range(range::Vector{Float64}, value::Number)
+    range[1] = min(range[1], abs(value))
+    range[2] = max(range[2], abs(value))
+    return 1
+end
+
+function _get_constraint_data(
+    data,
+    ref::ConstraintRef,
+    func::JuMP.GenericAffExpr,
+)
+    if length(func.terms) == 1
+        if first(values(func.terms)) ≈ 1.0
+            push!(data.bound_rows, ref)
+            data.matrix_nnz += 1
+            return
+        end
+    end
+    nnz = 0
+    for (variable, coefficient) in func.terms
+        if coefficient ≈ 0.0
```

Comment:
Note that this is exactly equivalent to `coefficient == 0.0`:
```julia
julia> isapprox(1e-20, 0.0)
false
```

We either mean `if iszero(coefficient)` or `≈(coefficient, 0.0; atol = 1e-12)` (or some other atol)

Reply:
Got it!
I copied that idea from SDDP.jl https://github.com/odow/SDDP.jl/blob/6d0e9a8a407266c7518687ecce8929cdc6bc9f94/src/print.jl#L278

I´d say we mean `iszero(coefficient)`, as we do want to tell user to be careful with `1e-12` reaching the solver.

Reply:
That code is from 2019! Old Oscar still had a bit to learn